In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
pd.set_option('display.max_columns', None)


#Data Cleaning and Exploration


df= pd.read_csv(r'C:\Users\Jethro Madusha\Downloads\July Amdari\Novapay-fraudlent\data\nova_pay_combined.csv',parse_dates=['timestamp' ])

df.head(5)



In [52]:

df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11400 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   transaction_id             11400 non-null  object 
 1   customer_id                11400 non-null  object 
 2   timestamp                  11371 non-null  object 
 3   home_country               11400 non-null  object 
 4   source_currency            11400 non-null  object 
 5   dest_currency              11400 non-null  object 
 6   channel                    11400 non-null  object 
 7   amount_src                 11400 non-null  object 
 8   amount_usd                 11095 non-null  float64
 9   fee                        11105 non-null  float64
 10  exchange_rate_src_to_dest  11400 non-null  float64
 11  device_id                  11400 non-null  object 
 12  new_device                 11400 non-null  bool   
 13  ip_address                 11095 non-null  obj

The dataset contains missing values across several variables, indicating that some transaction records are incomplete. The **amount_usd** and **ip_address** columns have the highest number of missing values, with **305 missing entries each**. Similarly, the **ip_country**, **kyc_tier**, **fee**, and **device_trust_score** columns also contain a substantial number of missing records. In comparison, the **timestamp** column has relatively few missing values, with only **29 missing entries**. These missing values should be appropriately handled during the data preprocessing stage to ensure the accuracy and reliability of subsequent analyses and predictive modeling.


# Visualize the missing data
plt.figure(figsize =(16,5))
sns.heatmap(df.isnull(), cbar=True, cmap="magma_r");

In [53]:
df.dropna(inplace=True)

df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 11066 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   transaction_id             11066 non-null  object 
 1   customer_id                11066 non-null  object 
 2   timestamp                  11066 non-null  object 
 3   home_country               11066 non-null  object 
 4   source_currency            11066 non-null  object 
 5   dest_currency              11066 non-null  object 
 6   channel                    11066 non-null  object 
 7   amount_src                 11066 non-null  object 
 8   amount_usd                 11066 non-null  float64
 9   fee                        11066 non-null  float64
 10  exchange_rate_src_to_dest  11066 non-null  float64
 11  device_id                  11066 non-null  object 
 12  new_device                 11066 non-null  bool   
 13  ip_address                 11066 non-null  object 



#The dataset contains 11,066 transaction records and 26 variables, providing a comprehensive set of information for fraud analysis. After data preprocessing, all columns have 11,066 non-null values, indicating that there are no missing values remaining in the dataset. The variables consist of different data types, including 12 text (string) variables, 7 decimal (float) variables, 5 integer variables, and 2 Boolean (True/False) variables, showing that the dataset contains a mix of categorical and numerical information.


In [55]:
df.drop_duplicates(inplace=True)
df.info()


<class 'pandas.core.frame.DataFrame'>
Index: 10871 entries, 0 to 11399
Data columns (total 26 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   transaction_id             10871 non-null  object 
 1   customer_id                10871 non-null  object 
 2   timestamp                  10871 non-null  object 
 3   home_country               10871 non-null  object 
 4   source_currency            10871 non-null  object 
 5   dest_currency              10871 non-null  object 
 6   channel                    10871 non-null  object 
 7   amount_src                 10871 non-null  object 
 8   amount_usd                 10871 non-null  float64
 9   fee                        10871 non-null  float64
 10  exchange_rate_src_to_dest  10871 non-null  float64
 11  device_id                  10871 non-null  object 
 12  new_device                 10871 non-null  bool   
 13  ip_address                 10871 non-null  object 


#The dataset was successfully cleaned and preprocessed, resulting in **10,871 complete transaction records** across **26 variables**. All variables contain valid entries for every remaining record, confirming that the dataset is free of missing values following preprocessing. The reduction in the number of records compared to the original dataset indicates that incomplete, inconsistent, or unsuitable observations were removed during the data cleaning process to improve data quality. The final dataset comprises a balanced mix of categorical, numerical, and Boolean variables, making it well-prepared for exploratory data analysis, feature engineering, and fraud detection model development.


In [ ]:
# Updated notebook
